File: backend/app/routes/announcements.py
import uuid

from flask import Blueprint, jsonify, make_response, request
from flask_jwt_extended import jwt_required, verify_jwt_in_request

from app.services.announcement_service import (
    broadcast_announcement,
    dismiss_announcement,
    list_active_banners,
    list_user_inbox,
    mark_announcement_read,
    pick_primary_banner,
)
from app.utils.auth_decorators import resolve_jwt_user

announcements_bp = Blueprint("announcements", __name__)

VALID_CONTEXTS = ("public", "customer", "staff")


@announcements_bp.route("/announcements/active", methods=["GET"])
def active_announcements():
    context = (request.args.get("context") or "public").strip().lower()
    if context not in VALID_CONTEXTS:
        return jsonify({"error": "Invalid context"}), 400

    user = None
    try:
        verify_jwt_in_request(optional=True)
        user, _ = resolve_jwt_user()
    except Exception:
        user = None

    banners = list_active_banners(context, user=user)
    primary = pick_primary_banner(banners)
    modals = [b for b in banners if b["display_as"] == "modal"]

    response = make_response(
        jsonify(
            {
                "banner": primary if primary and primary.get("display_as") == "banner" else None,
                "modals": modals,
            }
        )
    )
    response.headers["Cache-Control"] = "public, max-age=60"
    return response


@announcements_bp.route("/me/announcements", methods=["GET"])
@jwt_required()
def my_announcements():
    user, auth_err = resolve_jwt_user()
    if auth_err:
        return auth_err

    return jsonify({"announcements": list_user_inbox(user)})


@announcements_bp.route("/me/announcements/<announcement_id>/dismiss", methods=["POST"])
@jwt_required()
def dismiss_my_announcement(announcement_id: str):
    user, auth_err = resolve_jwt_user()
    if auth_err:
        return auth_err

    try:
        aid = uuid.UUID(announcement_id)
    except ValueError:
        return jsonify({"error": "Invalid announcement ID"}), 400

    try:
        dismiss_announcement(user, aid)
    except ValueError as exc:
        return jsonify({"error": str(exc)}), 400

    return jsonify({"message": "Dismissed"})


@announcements_bp.route("/me/announcements/<announcement_id>/read", methods=["POST"])
@jwt_required()
def read_my_announcement(announcement_id: str):
    user, auth_err = resolve_jwt_user()
    if auth_err:
        return auth_err

    try:
        aid = uuid.UUID(announcement_id)
    except ValueError:
        return jsonify({"error": "Invalid announcement ID"}), 400

    try:
        mark_announcement_read(user, aid)
    except ValueError as exc:
        return jsonify({"error": str(exc)}), 400

    return jsonify({"message": "Marked as read"})

File: server/src/routes/announcements.ts
import { Router } from 'express'
import { jwtOptional, jwtRequired, requireAuth, resolveJwtUser, type AuthRequest } from '../middleware/auth.js'
import { routeParam } from '../lib/routeParams.js'
import {
  dismissAnnouncement,
  listActiveBanners,
  listUserInbox,
  markAnnouncementRead,
  pickPrimaryBanner,
} from '../services/announcementService.js'

export const announcementsRouter = Router()

const VALID_CONTEXTS = new Set(['public', 'customer', 'staff'])

announcementsRouter.get('/announcements/active', jwtOptional, async (req: AuthRequest, res) => {
  const context = String(req.query.context ?? 'public')
    .trim()
    .toLowerCase()
  if (!VALID_CONTEXTS.has(context)) return res.status(400).json({ error: 'Invalid context' })

  let user = null
  if (req.jwtClaims) {
    try {
      user = await resolveJwtUser(req, res, { requireActive: true })
      if (!user && res.headersSent) return
    } catch {
      user = null
    }
  }

  const banners = await listActiveBanners(context, user)
  const primary = pickPrimaryBanner(banners)
  const modals = banners.filter((b) => b.display_as === 'modal')

  res.set('Cache-Control', 'public, max-age=60')
  return res.json({
    banner: primary && primary.display_as === 'banner' ? primary : null,
    modals,
  })
})

announcementsRouter.get('/me/announcements', jwtRequired, requireAuth, async (req: AuthRequest, res) => {
  return res.json({ announcements: await listUserInbox(req.user!) })
})

announcementsRouter.post('/me/announcements/:announcementId/dismiss', jwtRequired, requireAuth, async (req: AuthRequest, res) => {
  try {
    await dismissAnnouncement(req.user!, routeParam(req.params.announcementId))
    return res.json({ message: 'Dismissed' })
  } catch (exc) {
    return res.status(400).json({ error: exc instanceof Error ? exc.message : String(exc) })
  }
})

announcementsRouter.post('/me/announcements/:announcementId/read', jwtRequired, requireAuth, async (req: AuthRequest, res) => {
  try {
    await markAnnouncementRead(req.user!, routeParam(req.params.announcementId))
    return res.json({ message: 'Marked as read' })
  } catch (exc) {
    return res.status(400).json({ error: exc instanceof Error ? exc.message : String(exc) })
  }
})
